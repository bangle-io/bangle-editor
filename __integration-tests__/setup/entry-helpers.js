import React from 'react';
import reactDOM from 'react-dom';
import { ReactEditorView, useEditorState } from '@banglejs/react';
import { SpecRegistry } from '@banglejs/core/spec-registry';
import {
  defaultPlugins,
  defaultSpecs,
} from '@banglejs/core/test-helpers/default-components';

export function setupReactEditor({
  specRegistry,
  plugins = defaultPlugins,
  renderNodeViews,
  id = 'pm-root',
} = {}) {
  const element = document.createElement('div');
  window.document.body.appendChild(element);
  element.setAttribute('id', 'root');
  if (!(specRegistry instanceof SpecRegistry)) {
    specRegistry = new SpecRegistry(defaultSpecs(specRegistry));
  }
  reactDOM.render(
    <App opts={{ specRegistry, plugins, renderNodeViews, id }} />,
    element,
  );
}

function App({ opts: { specRegistry, plugins, renderNodeViews, id } }) {
  const onEditorReady = (_editor) => {
    window.editor = _editor;
  };

  window.dispatcher = (command) => {
    return command(
      window.editor.view.state,
      window.editor.view.dispatch,
      window.editor.view,
    );
  };

  const editorState = useEditorState({ specRegistry, plugins });

  return (
    <ReactEditorView
      id={id}
      editorState={editorState}
      onReady={onEditorReady}
      renderNodeViews={renderNodeViews}
    />
  );
}
