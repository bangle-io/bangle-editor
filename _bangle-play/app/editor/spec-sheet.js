import '@banglejs/core/style.css';
import '@banglejs/tooltip/style.css';

import * as collab from '@banglejs/collab/client/collab-extension';
import { emoji } from '@banglejs/emoji/index';
import '@banglejs/emoji/emoji.css';
import '@banglejs/react-menu/style.css';
import '@banglejs/markdown-front-matter/style.css';
import './extensions-override.css';
import { trailingNode } from '@banglejs/trailing-node';
import { timestamp } from '@banglejs/timestamp';
import { SpecRegistry } from '@banglejs/core/spec-registry';
import stopwatch from '@banglejs/react-stopwatch';
import sticker from '@banglejs/react-sticker';
import { emojiSuggestMenu } from '@banglejs/react-menu';
import { markdownFrontMatter } from '@banglejs/markdown-front-matter';
import {
  bold,
  code,
  italic,
  strike,
  link,
  underline,
  doc,
  text,
  paragraph,
  blockquote,
  bulletList,
  codeBlock,
  hardBreak,
  heading,
  horizontalRule,
  listItem,
  orderedList,
  todoItem,
  todoList,
  image,
} from '@banglejs/core/components/index';

export const specRegistry = new SpecRegistry([
  doc.spec({ content: 'frontMatter? block+' }),
  text.spec(),
  paragraph.spec(),
  blockquote.spec(),
  bulletList.spec(),
  codeBlock.spec(),
  hardBreak.spec(),
  heading.spec(),
  horizontalRule.spec(),
  listItem.spec(),
  orderedList.spec(),
  todoItem.spec(),
  todoList.spec(),
  image.spec(),
  bold.spec(),
  code.spec(),
  italic.spec(),
  strike.spec(),
  link.spec(),
  underline.spec(),
  collab.spec(),
  emoji.spec(),
  emojiSuggestMenu.spec(),
  stopwatch.spec(),
  trailingNode.spec(),
  timestamp.spec(),
  sticker.spec(),
  markdownFrontMatter.spec(),
]);
