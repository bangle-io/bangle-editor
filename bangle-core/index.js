import * as logging from './utils/logging';
import browser, { isChromeWithSelectionBug } from './utils/browser';
export * from './components/index';
export * from './editor';
export { logging, browser, isChromeWithSelectionBug };
export * from './pm-plugin';
export * from './node-view';
export * from './pm-model';
