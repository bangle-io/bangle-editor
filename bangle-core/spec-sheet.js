import { Schema } from 'prosemirror-model';
import { coreSpec } from './components/index';
import { bangleWarn, recursiveFlat } from './utils/js-utils';

export class SpecSheet {
  constructor(rawSpecs = coreSpec()) {
    this._spec = recursiveFlat(rawSpecs);

    this._spec.forEach(validateSpec);

    if (this._spec.length !== new Set(this._spec.map((r) => r.name)).size) {
      bangleWarn(
        'The specSheet has one or more specs with the same name',
        this._spec,
      );
      throw new Error('Duplicate spec error, please check your specSheet');
    }
    this._schema = createSchema(this._spec);
    this._options = Object.fromEntries(
      this._spec
        .filter((spec) => spec.options)
        .map((spec) => [spec.name, spec.options]),
    );
  }

  get spec() {
    return this._spec;
  }

  get schema() {
    return this._schema;
  }

  get options() {
    return this._options;
  }
}

function createSchema(specSheet) {
  let nodes = [];
  let marks = [];
  let topNode;
  for (const spec of specSheet) {
    if (spec.type === 'node') {
      nodes.push([spec.name, spec.schema]);
    } else if (spec.type === 'mark') {
      marks.push([spec.name, spec.schema]);
    } else if (spec.type === 'component') {
    } else {
      throw new Error(spec.name + ' unknown type: ' + spec.type);
    }
    if (spec.topNode === true) {
      topNode = spec.name;
    }
  }

  return new Schema({
    topNode,
    nodes: Object.fromEntries(nodes),
    marks: Object.fromEntries(marks),
  });
}

function validateSpec(spec) {
  if (!spec.name) {
    bangleWarn("The spec didn't have a name field", spec);
    throw new Error('Spec must have a name');
  }
  if (!['node', 'mark', 'component'].includes(spec.type)) {
    bangleWarn('The spec must be of type node, mark or component ', spec);
    throw new Error('Invalid spec type');
  }
  if (['node', 'mark'].includes(spec.type) && !spec.schema) {
    bangleWarn(
      "The spec of type 'mark' or 'node' must have a schema field",
      spec,
    );
    throw new Error('Invalid spec schema');
  }
}
