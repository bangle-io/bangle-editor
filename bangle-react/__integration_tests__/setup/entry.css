@import 'bangle-core/style.css';
