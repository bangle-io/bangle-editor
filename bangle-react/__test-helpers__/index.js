export * from './pjsx';
export * from './react-test-editor';
