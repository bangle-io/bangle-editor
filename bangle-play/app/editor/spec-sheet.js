import 'bangle-core/style.css';
import { coreSpec } from 'bangle-core/components';
import * as collab from 'bangle-plugins/collab/client/collab-extension';
import { emoji, emojiInlineSuggest } from 'bangle-plugins/emoji/index';
import 'bangle-plugins/emoji/emoji.css';
// import 'bangle-react/inline-menu/inline-menu.css';
import 'bangle-react/menu/style.css';

import { config } from './config';
import { trailingNode } from 'bangle-plugins/trailing-node/index';
import { timestamp } from 'bangle-plugins/timestamp/index';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { dino } from 'bangle-react/dino';
import { stopwatch } from 'bangle-react/stopwatch';

export const specSheet = new SpecSheet([
  ...coreSpec({ heading: { levels: config.headingLevels } }),
  collab.spec(),
  emoji.spec(),
  emojiInlineSuggest.spec({ markName: 'emoji_inline_suggest', trigger: ':' }),
  stopwatch.spec(),
  trailingNode.spec(),
  timestamp.spec(),
  dino.spec(),
]);
