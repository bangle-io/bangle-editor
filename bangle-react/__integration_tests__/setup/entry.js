import './entry.css';
import React from 'react';
import reactDOM from 'react-dom';
import { corePlugins, coreSpec } from 'bangle-core/index';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { Slice } from 'prosemirror-model';
import { DOMSerializer } from 'prosemirror-model';
import { dino } from 'bangle-react/dino';
import { stopwatch } from 'bangle-react/stopwatch';
import * as prosemirrorView from 'prosemirror-view';
import { ReactEditor } from 'bangle-react/ReactEditor';

window.prosemirrorView = prosemirrorView;

console.debug('Bangle-react entry.js');
setup();

function setup() {
  const element = document.createElement('div');
  window.document.body.appendChild(element);
  element.setAttribute('id', 'root');
  // const editorContainer = document.createElement('div');
  // editorContainer.setAttribute('id', 'pm-root');
  // element.appendChild(editorContainer);
  // setupEditor(editorContainer);
  reactDOM.render(<App />, element);
}

function App() {
  const specSheet = new SpecSheet([
    ...coreSpec(),
    stopwatch.spec(),
    dino.spec(),
  ]);
  const plugins = [...corePlugins(), stopwatch.plugins(), dino.plugins()];
  window.commands = {
    stopwatch: stopwatch.commands,
    dino: dino.commands,
  };
  const onEditorReady = (_editor) => {
    window.editor = _editor;
  };
  const renderNodeViews = ({ node, ...args }) => {
    if (node.type.name === 'dino') {
      return <dino.Dino node={node} {...args} />;
    }
  };
  window.dispatcher = (command) => {
    return command(
      window.editor.view.state,
      window.editor.view.dispatch,
      window.editor.view,
    );
  };
  return (
    <ReactEditor
      options={{ id: 'pm-root', specSheet, plugins }}
      onReady={onEditorReady}
      renderNodeViews={renderNodeViews}
    />
  );
}

const createEvent = (name, options = {}) => {
  let event;
  if (options.bubbles === undefined) {
    options.bubbles = true;
  }
  if (options.cancelable === undefined) {
    options.cancelable = true;
  }
  if (options.composed === undefined) {
    options.composed = true;
  }
  event = new Event(name, options);

  return event;
};

window.manualPaste = function manualPaste(htmlStr) {
  function sliceSingleNode(slice) {
    return slice.openStart === 0 &&
      slice.openEnd === 0 &&
      slice.content.childCount === 1
      ? slice.content.firstChild
      : null;
  }

  function doPaste(view, text, html, e) {
    let slice = prosemirrorView.__parseFromClipboard(
      view,
      text,
      html,
      view.shiftKey,
      view.state.selection.$from,
    );

    if (
      view.someProp('handlePaste', (f) => f(view, e, slice || Slice.empty)) ||
      !slice
    ) {
      return;
    }

    let singleNode = sliceSingleNode(slice);
    let tr = singleNode
      ? view.state.tr.replaceSelectionWith(singleNode, view.shiftKey)
      : view.state.tr.replaceSelection(slice);
    view.dispatch(
      tr.scrollIntoView().setMeta('paste', true).setMeta('uiEvent', 'paste'),
    );
  }
  const pasteEvent = createEvent('paste');

  doPaste(window.editor.view, null, htmlStr, pasteEvent);
};

window.domSerializer = function domSerializer() {
  return DOMSerializer.fromSchema(window.editor.view.state.schema);
};
