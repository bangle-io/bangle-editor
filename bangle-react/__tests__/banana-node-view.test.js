/** @jsx pjsx */

import { screen } from '@testing-library/dom';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { corePlugins, coreSpec } from 'bangle-core/index';
import { pjsx, reactTestEditor } from 'bangle-react/__test-helpers__/index';
import { bananaComponent, Banana } from './setup/banana';

const renderNodeViews = jest.fn(({ node, ...args }) => {
  if (node.type.name === 'banana') {
    return <Banana node={node} {...args} />;
  }
  throw new Error('Unknown node');
});

describe('Inline node banana', () => {
  test('Inits banana', async () => {
    const banana = bananaComponent();
    const specSheet = new SpecSheet([...coreSpec(), banana.spec()]);
    const plugins = [...corePlugins(), banana.plugins()];

    const testEditor = reactTestEditor({ specSheet, plugins, renderNodeViews });

    const { view, editor } = await testEditor(
      <doc>
        <heading>Wow[]</heading>
        <para>
          Child <banana />
        </para>
      </doc>,
    );

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <heading>Wow[]</heading>
        <para>
          Child <banana />
        </para>
      </doc>,
    );

    expect(editor.toHTMLString()).toMatchSnapshot();
  });

  test('Can update attrs', async () => {
    const testId = 'Can update attrs';
    const banana = bananaComponent(testId);
    const specSheet = new SpecSheet([...coreSpec(), banana.spec()]);
    const plugins = [...corePlugins(), banana.plugins()];
    const testEditor = reactTestEditor({ specSheet, plugins, renderNodeViews });

    const { view, posLabels } = await testEditor(
      <doc>
        <heading>Wow</heading>
        <para>
          Child <banana />
          []
        </para>
      </doc>,
    );

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <heading>Wow</heading>
        <para>
          Child <banana />
          []
        </para>
      </doc>,
    );

    const el = await screen.findByTestId(testId);
    await expect(el).toMatchInlineSnapshot(`
            <span
              data-bangle-container=""
              data-bangle-name="banana"
              data-testid="Can update attrs"
              draggable="true"
            >
              <div
                data-color="yellow"
              >
                I am fresh yellow banana
              </div>
            </span>
          `);
    const bananaNodePos = posLabels['[]'] - 1;

    view.dispatch(
      view.state.tr.setNodeMarkup(bananaNodePos, undefined, {
        color: 'brown',
      }),
    );

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <heading>Wow</heading>
        <para>
          Child <banana color="brown" />
          []
        </para>
      </doc>,
    );

    await expect(el).toMatchInlineSnapshot(`
            <span
              data-bangle-container=""
              data-bangle-name="banana"
              data-testid="Can update attrs"
              draggable="true"
            >
              <div
                data-color="brown"
              >
                I am fresh brown banana
              </div>
            </span>
          `);
  });
});
