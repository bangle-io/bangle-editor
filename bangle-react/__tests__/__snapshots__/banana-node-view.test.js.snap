// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline node banana Inits banana 1`] = `"<h1>Wow</h1><p>Child <span data-bangle-name=\\"banana\\" data-bangle-attrs=\\"{&quot;ripe&quot;:&quot;fresh&quot;,&quot;color&quot;:&quot;yellow&quot;}\\"></span></p>"`;
